File: python3/Trees/invert_binary_tree_recursive.py
from ds import TreeNode

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
"""


def invert_binary_tree_recursive(root: TreeNode) -> TreeNode:
    # Base case: If the node is null, there's nothing to invert.
    if not root:
        return None
    # Swap the left and right subtrees of the current node.
    root.left, root.right = root.right, root.left
    # Recursively invert the left and right subtrees.
    invert_binary_tree_recursive(root.left)
    invert_binary_tree_recursive(root.right)
    return root

File: kotlin/Trees/InvertBinaryTreeRecursive.kt
import ds.TreeNode

/*
    Definition of TreeNode:

    data class TreeNode(
        var value: Int, 
        var left: TreeNode? = null, 
        var right: TreeNode? = null
    )
*/

fun invertBinaryTreeRecursive(root: TreeNode?): TreeNode? {
    // Base case: If the node is null, there's nothing to invert.
    if (root == null) {
        return null
    }
    // Swap the left and right subtrees of the current node.
    val temp = root.left
    root.left = root.right
    root.right = temp
    // Recursively invert the left and right subtrees.
    invertBinaryTreeRecursive(root.left)
    invertBinaryTreeRecursive(root.right)
    return root
}
